File: manual/src/test/java/RetrofitDocTest.java
import static org.assertj.core.api.Assertions.assertThat;

import java.io.IOException;
import java.util.List;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.parallel.Execution;
import org.junit.jupiter.api.parallel.ExecutionMode;
import org.shakespeareframework.Actor;
import org.shakespeareframework.Question;
import org.shakespeareframework.retrofit.CallHttpApis;
import retrofit2.Call;
import retrofit2.http.GET;
import retrofit2.http.Path;

@Execution(ExecutionMode.SAME_THREAD)
class RetrofitDocTest {

  static MockWebServer serviceMock = new MockWebServer();
  static String ACTORS_SERVICE_URL = serviceMock.url("/").toString();

  @AfterAll
  static void shutdownMockWebServer() throws IOException {
    serviceMock.shutdown();
  }

  // tag::test-api[]
  interface ActorsApi {

    @GET("/actors")
    Call<List<ActorInfo>> getActors();

    @GET("/actors/{id}/name")
    Call<String> getActorName(@Path("id") String id);
  }
  // end::test-api[]

  // tag::scalars-example[]
  record ActorName(String id) implements Question<String> {

    @Override
    public String answerAs(Actor actor) {
      var actorsApiClient =
          actor
              .uses(CallHttpApis.class)
              .buildClient() // <1>
              .baseUrl(ACTORS_SERVICE_URL) // <2>
              .addScalarsConverterFactory() // <3>
              .build(ActorsApi.class); // <4>
      try {
        return actorsApiClient.getActorName(id).execute().body();
      } catch (IOException e) {
        throw new RuntimeException("%s failed".formatted(this), e);
      }
    }
  }
  // end::scalars-example[]

  @Test
  void act1() {
    Actor ron = new Actor("Ron").can(new CallHttpApis());
    var jsonBody =
        """
                [{"id":"273","name":"Alex"},{"id":"476","name":"Jonny"},{"id":"538","name":"Emma"}]
                """;
    serviceMock.enqueue(new MockResponse().setBody(jsonBody));

    var actors = ron.checks(new AllActors());

    assertThat(actors).map(ActorInfo::name).containsExactly("Alex", "Jonny", "Emma");
  }

  @Test
  void act2() {
    Actor scarlet = new Actor("Scarlet").can(new CallHttpApis());
    serviceMock.enqueue(new MockResponse().setBody("Alex"));

    var actors = scarlet.checks(new ActorName("273"));

    assertThat(actors).isEqualTo("Alex");
  }

  // tag::jackson-example[]
  record AllActors() implements Question<List<ActorInfo>> {

    @Override
    public List<ActorInfo> answerAs(Actor actor) {
      var actorsApiClient =
          actor
              .uses(CallHttpApis.class)
              .buildClient()
              .baseUrl(ACTORS_SERVICE_URL)
              .addJacksonConverterFactory() // <1>
              .build(ActorsApi.class);
      try {
        return actorsApiClient.getActors().execute().body();
      } catch (IOException e) {
        throw new RuntimeException("%s failed".formatted(this), e);
      }
    }
  }

  record ActorInfo(String id, String name) {} // <2>
  // end::jackson-example[]
}

File: manual/src/test/kotlin/examples/kotlin/RetrofitDocTest.kt
package examples.kotlin

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.kotlinModule
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.parallel.Execution
import org.junit.jupiter.api.parallel.ExecutionMode
import org.shakespeareframework.Actor
import org.shakespeareframework.Question
import org.shakespeareframework.retrofit.CallHttpApis
import retrofit2.Call
import retrofit2.http.GET
import retrofit2.http.Path
import java.io.IOException

@Execution(ExecutionMode.SAME_THREAD)
class RetrofitDocTest {

  companion object {
    val serviceMock = MockWebServer()
    val serviceUrl = serviceMock.url("/").toString()

    @JvmStatic
    @AfterAll
    @Throws(IOException::class)
    fun shutdownMockWebServer() {
      serviceMock.shutdown()
    }
  }

  // tag::test-api[]
  interface ActorsApi {
    @GET("/actors")
    fun getActors(): Call<List<ActorInfo>>

    @GET("/actors/{id}/name")
    fun getActorName(@Path("id") id: String): Call<String>
  }
  // end::test-api[]

  // tag::scalars-example[]
  data class ActorName(val id: String) : Question<String> {
    override fun answerAs(actor: Actor): String {
      val actorsApiClient = actor
          .uses(CallHttpApis::class.java)
                .buildClient() // <1>
          .baseUrl(serviceUrl) // <2>
          .addScalarsConverterFactory() // <3>
          .build(ActorsApi::class.java) // <4>
      return try {
        actorsApiClient.getActorName(id).execute().body() ?: throw RuntimeException("No body")
      } catch (e: IOException) {
        throw RuntimeException("${this} failed", e)
      }
    }
  }
  // end::scalars-example[]

  @Test
  @Throws(IOException::class)
  fun act1() {
    val ron = Actor("Ron").can(CallHttpApis())
    val jsonBody = """
            [{"id":"273","name":"Alex"},{"id":"476","name":"Jonny"},{"id":"538","name":"Emma"}]
        """
    serviceMock.enqueue(MockResponse().setBody(jsonBody))

    val actors = ron.checks(AllActors())

    assertThat(actors)
      .flatMap(ActorInfo::name)
      .containsExactly("Alex", "Jonny", "Emma")
  }

  @Test
  @Throws(IOException::class)
  fun act2() {
    val scarlet = Actor("Scarlet").can(CallHttpApis())
    serviceMock.enqueue(MockResponse().setBody("Alex"))

    val actorName = scarlet.checks(ActorName("273"))

    assertThat(actorName).isEqualTo("Alex")
  }

  // tag::jackson-example[]
  class AllActors : Question<List<ActorInfo>> {
    override fun answerAs(actor: Actor): List<ActorInfo> {
      val actorsApiClient = actor
          .uses(CallHttpApis::class.java)
                .buildClient()
          .baseUrl(serviceUrl)
          .addJacksonConverterFactory(ObjectMapper().registerModule(kotlinModule())) // <1>
          .build(ActorsApi::class.java)
      return try {
        actorsApiClient.getActors().execute().body() ?: throw RuntimeException("No body")
      } catch (e: IOException) {
        throw RuntimeException("${this} failed", e)
      }
    }
  }

  data class ActorInfo(val id: String, val name: String) // <2>
  // end::jackson-example[]
}
